### PR TITLE
Add NodeRequirement

### PR DESCRIPTION
**Pull request by henriquegemignani**
Adds the option for a requirement in the database to require collecting one specific node. The UI lets one select any node, but it only makes sense if the node is a resource node.

I intend to use this feature for things in the future :)

### CODE REVIEW

File: randovania/gui/dialog/connections_editor.py [hunk 1]
```diff
@@ -92,7 +95,22 @@ def _create_default_template_requirement(resource_database: ResourceDatabase) ->
     return RequirementTemplate(template_name)
 
 
-class ResourceRequirementEditor:
+def _create_default_node_requirement(region_list: RegionList) -> NodeRequirement:
+    for node in region_list.all_nodes:
+        return NodeRequirement(node.identifier)
```
Comment (Miepee) [inline randovania/gui/dialog/connections_editor.py:100]: mypy

File: randovania/gui/dialog/connections_editor.py [hunk 2]
```diff
@@ -92,7 +95,22 @@ def _create_default_template_requirement(resource_database: ResourceDatabase) ->
...
+    raise RuntimeError("No nodes?!")
+
+
+class BaseEditor:
+    def deleteLater(self):
```
Comment (Miepee) [inline randovania/gui/dialog/connections_editor.py:105]: mypy

File: randovania/gui/dialog/connections_editor.py [hunk 3]
```diff
@@ -308,15 +330,45 @@ def current_requirement(self) -> RequirementTemplate:
...
+        item: NodeRequirement,
+    ):
+        self.parent = parent
+        self.layout = layout
+        self.selector = NodeSelectorWidget(region_list)
+        self.selector.select_by_identifier(item.node_identifier)
+
+        self.layout.addWidget(self.selector)
+
+    def deleteLater(self):
```
Comment (Miepee) [inline randovania/gui/dialog/connections_editor.py:348]: mypy

File: randovania/gui/docks/resource_database_editor.py [hunk 4]
```diff
@@ -293,7 +293,7 @@ def rename_template(self, name: str, display_name: str) -> None:
 
     def edit_template(self, name: str) -> None:
         template = self.db.requirement_template[name]
-        editor = ConnectionsEditor(self, self.db, template.requirement)
+        editor = ConnectionsEditor(self, self.db, self.region_list, template.requirement)
```
Comment (Miepee) [inline randovania/gui/docks/resource_database_editor.py:296]: mypy

File: randovania/gui/lib/signal_handling.py [hunk 5]
```diff
@@ -35,3 +35,15 @@ def set_combo_with_value(combo: QtWidgets.QComboBox, value):
     :return:
     """
     combo.setCurrentIndex(combo.findData(value))
+
+
+def refresh_if_needed(combo: QtWidgets.QComboBox, func) -> None:
```
Comment (Miepee) [inline randovania/gui/lib/signal_handling.py:40]: mypy

File: randovania/gui/widgets/node_selector_widget.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        signal_handling.clear_without_notify(self.node_combo)
+        for node in area.nodes:
+            self.node_combo.addItem(node.name, userData=node)
+        assert self.node_combo.count() > 0
+
+    def on_node_combo(self, _: None) -> None:
+        self.SelectedNodeChanged.emit()
+
+    def select_by_identifier(self, identifier: NodeIdentifier) -> None:
+        self.region_combo.setCurrentIndex(self.region_combo.findText(identifier.region_name))
```
Comment (Miepee) [inline randovania/gui/widgets/node_selector_widget.py:64]: mypy (and below due to same cause)

File: test/gui/dialog/test_connections_editor.py [hunk 7]
```diff
@@ -25,7 +27,7 @@ def mk_req(name: str):
     )
 
     # Run
-    editor = connections_editor.ConnectionsEditor(None, echoes_resource_database, requirement)
+    editor = connections_editor.ConnectionsEditor(None, gd.resource_database, gd.region_list, requirement)
```
Comment (Miepee) [inline test/gui/dialog/test_connections_editor.py:30]: mypy, probably needs the # type-ignore

File: randovania/game_description/pretty_print.py [hunk 8]
```diff
@@ -94,6 +99,9 @@ def pretty_format_requirement(
 
     elif isinstance(requirement, RequirementTemplate):
         yield level, get_template_name(db, requirement)
+
+    elif isinstance(requirement, NodeRequirement):
+        yield level, requirement.node_identifier.as_string
```
Comment (Miepee) [inline randovania/game_description/pretty_print.py:104]: IMO displaying just the node identifier is a tad confusing:
Maybe have `Collected` before it?

**Review by Miepee — COMMENTED** [review]
If you have add a node requirement to region/area/my node , and then rename `my node`, saving the database throws. Would be nice if that could get handled, since renaming nodes happens somewhat regularly. 
(FWIW I only care about nodes, not region/area renames.)

**Comment by Miepee** [comment]
Other things I found on this branch. *Pretty* sure they're not in main, but havent actually confirmed:
- If you create a node via "Create Dock to this Area" it will error out on every node edit operation with this stacktrace:
```
Traceback (most recent call last):
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/gui/data_editor.py", line 669, in _move_node
    self.replace_node_with(area, node, dataclasses.replace(node, location=location))
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/gui/data_editor.py", line 407, in replace_node_with
    self.editor.replace_node(area, old_node, new_node)
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/editor.py", line 100, in replace_node
    self.remove_node(area, old_node.lock_node)
                           ^^^^^^^^^^^^^^^^^^
AttributeError: 'DockNode' object has no attribute 'lock_node'

```


- If you have a node connection on a lock of a dock node, then renaming the dock node and saving fails with this stacktrace:
```
Traceback (most recent call last):
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/gui/data_editor.py", line 645, in _save_as_internal_database
    if self._save_database(self._data_path):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/gui/data_editor.py", line 635, in _save_database
    data = data_writer.write_game_description(self.game_description)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/data_writer.py", line 446, in write_game_description
    "used_trick_levels": write_used_trick_levels(game),
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/data_writer.py", line 429, in write_used_trick_levels
    trick.short_name: sorted(levels) for trick, levels in game.get_used_trick_levels(ignore_cache=True).items()
                                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/game_description.py", line 183, in get_used_trick_levels
    process(requirement)
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/game_description.py", line 170, in process
    for resource_requirement in req.iterate_resource_requirements(context):
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/requirements/array_base.py", line 67, in iterate_resource_requirements
    yield from item.iterate_resource_requirements(context)
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/requirements/node_requirement.py", line 77, in iterate_resource_requirements
    yield self.as_resource_requirement(context)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/requirements/node_requirement.py", line 68, in as_resource_requirement
    return ResourceRequirement.simple(NodeResourceInfo.from_identifier(self.node_identifier, context))
                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/resources/node_resource_info.py", line 38, in from_identifier
    return cls.from_node(context.node_provider.node_by_identifier(identifier), context)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/db/region_list.py", line 247, in node_by_identifier
    raise ValueError(f"No node with name {identifier.node} found in {area}")
ValueError: No node with name Lock - Door to Surface Stash found in Area[Landing Site]
```

**Comment by Miepee** [comment]
- Renaming an Area now completely throws:
```
Traceback (most recent call last):
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/gui/data_editor.py", line 645, in _save_as_internal_database
    if self._save_database(self._data_path):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/gui/data_editor.py", line 630, in _save_database
    errors = integrity_check.find_database_errors(self.game_description)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/integrity_check.py", line 277, in find_database_errors
    result.extend(find_invalid_strongly_connected_components(game))
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/integrity_check.py", line 215, in find_invalid_strongly_connected_components
    starting_node = game.region_list.node_by_identifier(game.starting_location)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/db/region_list.py", line 241, in node_by_identifier
    area = self.area_by_area_location(identifier.area_identifier)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/db/region_list.py", line 258, in area_by_area_location
    return self.region_and_area_by_area_identifier(location)[1]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/db/region_list.py", line 265, in region_and_area_by_area_identifier
    area = region.area_by_name(identifier.area)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/db/region.py", line 48, in area_by_name
    raise KeyError(f"Unknown name: {area_name}")
KeyError: 'Unknown name: Landing Site'
```

**Comment by henriquegemignani** [comment]
>     * If you create a node via "Create Dock to this Area" it will error out on every node edit operation with this stacktrace:

This one happens in main.

**Comment by henriquegemignani** [comment]
> * Renaming an Area now completely throws:
> 
> 
> ```
> Traceback (most recent call last):
>   File "/home/narr/Dokumente/gitRepos/randovania/randovania/gui/data_editor.py", line 645, in _save_as_internal_database
>     if self._save_database(self._data_path):
>        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/narr/Dokumente/gitRepos/randovania/randovania/gui/data_editor.py", line 630, in _save_database
>     errors = integrity_check.find_database_errors(self.game_description)
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/integrity_check.py", line 277, in find_database_errors
>     result.extend(find_invalid_strongly_connected_components(game))
>   File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/integrity_check.py", line 215, in find_invalid_strongly_connected_components
>     starting_node = game.region_list.node_by_identifier(game.starting_location)
>                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/db/region_list.py", line 241, in node_by_identifier
>     area = self.area_by_area_location(identifier.area_identifier)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/db/region_list.py", line 258, in area_by_area_location
>     return self.region_and_area_by_area_identifier(location)[1]
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/db/region_list.py", line 265, in region_and_area_by_area_identifier
>     area = region.area_by_name(identifier.area)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/narr/Dokumente/gitRepos/randovania/randovania/game_description/db/region.py", line 48, in area_by_name
>     raise KeyError(f"Unknown name: {area_name}")
> KeyError: 'Unknown name: Landing Site'
> ```

Happens in main. It's related to renaming the area that is references in `GameDescription.starting_location`.

**Comment by henriquegemignani** [comment]
>     * If you have a node connection on a lock of a dock node, then renaming the dock node and saving fails with this stacktrace:

Fixed.

**Pull request merged**